File: src/main/java/szewek/flux/block/MachineBlock.java
package szewek.flux.block;

import net.minecraft.block.*;
import net.minecraft.block.material.Material;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.inventory.IInventory;
import net.minecraft.inventory.InventoryHelper;
import net.minecraft.inventory.container.Container;
import net.minecraft.inventory.container.INamedContainerProvider;
import net.minecraft.item.BlockItemUseContext;
import net.minecraft.item.ItemStack;
import net.minecraft.state.BooleanProperty;
import net.minecraft.state.DirectionProperty;
import net.minecraft.state.StateContainer;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.tileentity.TileEntityType;
import net.minecraft.util.*;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.BlockRayTraceResult;
import net.minecraft.world.IBlockReader;
import net.minecraft.world.World;
import net.minecraftforge.registries.ForgeRegistries;
import szewek.flux.tile.AbstractMachineTile;

import javax.annotation.Nullable;

public final class MachineBlock extends Block {
	public static final DirectionProperty FACING = HorizontalBlock.FACING;
	public static final BooleanProperty LIT = RedstoneTorchBlock.LIT;

	public MachineBlock() {
		super(Properties.of(Material.METAL)
				.strength(1.0F)
				.sound(SoundType.METAL)
				.lightLevel(state -> state.getValue(LIT) ? 13 : 0)
		);
		registerDefaultState(stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(LIT, false));
	}

	@Override
	public boolean hasTileEntity(BlockState state) {
		return true;
	}

	@Nullable
	@Override
	public TileEntity createTileEntity(BlockState state, IBlockReader world) {
		TileEntityType<?> tetype = ForgeRegistries.TILE_ENTITIES.getValue(getRegistryName());
		return tetype == null ? null : tetype.create();
	}

	@Override
	public ActionResultType use(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockRayTraceResult rayTrace) {
		if (!world.isClientSide) {
			TileEntity te = world.getBlockEntity(pos);
			if (te != null) {
				TileEntityType<?> type = ForgeRegistries.TILE_ENTITIES.getValue(this.getRegistryName());
				if (type == te.getType()) {
					player.openMenu((INamedContainerProvider) te);
				}
			}
		}
		return ActionResultType.SUCCESS;
	}

	@Override
	public BlockState getStateForPlacement(BlockItemUseContext context) {
		return defaultBlockState().setValue(FACING, context.getHorizontalDirection().getOpposite());
	}

	@Override
	public void setPlacedBy(World worldIn, BlockPos pos, BlockState state, @Nullable LivingEntity placer, ItemStack stack) {
		if (stack.hasCustomHoverName()) {
			TileEntity tileentity = worldIn.getBlockEntity(pos);
			if (tileentity instanceof AbstractMachineTile) {
				((AbstractMachineTile)tileentity).setCustomName(stack.getDisplayName());
			}
		}
	}

	@Override
	public void onRemove(BlockState state, World worldIn, BlockPos pos, BlockState newState, boolean isMoving) {
		if (state.getBlock() != newState.getBlock()) {
			TileEntity tileentity = worldIn.getBlockEntity(pos);
			if (tileentity instanceof AbstractMachineTile) {
				InventoryHelper.dropContents(worldIn, pos, (IInventory)tileentity);
				worldIn.updateNeighbourForOutputSignal(pos, this);
			}

			super.onRemove(state, worldIn, pos, newState, isMoving);
		}
	}

	@Override
	public boolean hasAnalogOutputSignal(BlockState state) {
		return true;
	}

	@Override
	public int getAnalogOutputSignal(BlockState blockState, World worldIn, BlockPos pos) {
		return Container.getRedstoneSignalFromBlockEntity(worldIn.getBlockEntity(pos));
	}

	@Override
	public BlockState rotate(BlockState state, Rotation rot) {
		return state.setValue(FACING, rot.rotate(state.getValue(FACING)));
	}

	@Override
	public BlockState mirror(BlockState state, Mirror mirrorIn) {
		return state.rotate(mirrorIn.getRotation(state.getValue(FACING)));
	}

	@Override
	protected void createBlockStateDefinition(StateContainer.Builder<Block, BlockState> builder) {
		builder.add(FACING, LIT);
	}
}

File: src/main/java/szewek/flux/block/MachineBlock.kt
package szewek.flux.block

import net.minecraft.block.*
import net.minecraft.block.material.Material
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.IInventory
import net.minecraft.inventory.InventoryHelper
import net.minecraft.inventory.container.Container
import net.minecraft.inventory.container.INamedContainerProvider
import net.minecraft.item.BlockItemUseContext
import net.minecraft.item.ItemStack
import net.minecraft.state.BooleanProperty
import net.minecraft.state.DirectionProperty
import net.minecraft.state.StateContainer
import net.minecraft.tileentity.TileEntity
import net.minecraft.tileentity.TileEntityType
import net.minecraft.util.*
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.BlockRayTraceResult
import net.minecraft.world.IBlockReader
import net.minecraft.world.World
import net.minecraftforge.registries.ForgeRegistries
import szewek.flux.tile.AbstractMachineTile

class MachineBlock : ContainerBlock(Properties.create(Material.IRON).hardnessAndResistance(1f).sound(SoundType.METAL).lightValue(13)) {
    override fun createNewTileEntity(worldIn: IBlockReader): TileEntity? {
        return ForgeRegistries.TILE_ENTITIES.getValue(registryName)?.create()
    }

    @Suppress("DEPRECATION")
    override fun getLightValue(state: BlockState): Int {
        return if (state[LIT]) super.getLightValue(state) else 0
    }

    override fun func_225533_a_(state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: Hand, rayTrace: BlockRayTraceResult): ActionResultType {
        if (world.isRemote) return ActionResultType.SUCCESS
        val te = world.getTileEntity(pos)
        if (te != null) {
            val type: TileEntityType<*>? = ForgeRegistries.TILE_ENTITIES.getValue(registryName)
            if (type === te.type) {
                player.openContainer(te as INamedContainerProvider?)
            }
        }
        return ActionResultType.SUCCESS
    }

    override fun getStateForPlacement(context: BlockItemUseContext): BlockState? {
        return defaultState.with(FACING, context.placementHorizontalFacing.opposite)
    }

    override fun onBlockPlacedBy(worldIn: World, pos: BlockPos, state: BlockState, placer: LivingEntity?, stack: ItemStack) {
        if (stack.hasDisplayName()) {
            val tileentity = worldIn.getTileEntity(pos)
            if (tileentity is AbstractMachineTile) {
                tileentity.customName = stack.displayName
            }
        }
    }

    @Suppress("DEPRECATION")
    override fun onReplaced(state: BlockState, worldIn: World, pos: BlockPos, newState: BlockState, isMoving: Boolean) {
        if (state.block !== newState.block) {
            val tileentity = worldIn.getTileEntity(pos)
            if (tileentity is AbstractMachineTile) {
                InventoryHelper.dropInventoryItems(worldIn, pos, tileentity as IInventory)
                worldIn.updateComparatorOutputLevel(pos, this)
            }
            super.onReplaced(state, worldIn, pos, newState, isMoving)
        }
    }

    override fun hasComparatorInputOverride(state: BlockState) = true

    override fun getComparatorInputOverride(blockState: BlockState, worldIn: World, pos: BlockPos): Int {
        return Container.calcRedstone(worldIn.getTileEntity(pos))
    }

    override fun getRenderType(state: BlockState) = BlockRenderType.MODEL

    override fun rotate(state: BlockState, rot: Rotation): BlockState {
        return state.with(FACING, rot.rotate(state[FACING]))
    }

    override fun mirror(state: BlockState, mirrorIn: Mirror): BlockState {
        return state.rotate(mirrorIn.toRotation(state[FACING]))
    }

    override fun fillStateContainer(builder: StateContainer.Builder<Block, BlockState>) {
        builder.add(FACING, LIT)
    }

    companion object {
        val FACING: DirectionProperty = HorizontalBlock.HORIZONTAL_FACING
        val LIT: BooleanProperty = RedstoneTorchBlock.LIT
    }

    init {
        defaultState = stateContainer.baseState.with(FACING, Direction.NORTH).with(LIT, false)
    }
}